cmd/block-generator/generator/generate.go

```diff
@@ -290,35 +371,193 @@ func (g *generator) generatePaymentTxn(round uint64, intra uint64) (types.Signed
 }
 
-func (g *generator) generateAssetTxn(round uint64, intra uint64) (types.SignedTxnWithAD, error) {
-	return convert(sdk_types.Transaction{}, types.ApplyData{}), nil
+func getAssetTxOptions() []interface{} {
+	return []interface{}{assetCreate, assetDestroy, assetOptin, assetXfer, assetClose}
+}
+
+func (g *generator) generateAssetTxnInternal(txType interface{}, sp sdk_types.SuggestedParams, intra uint64) (txn types.Transaction) {
+	return g.generateAssetTxnInternalHint(txType, sp, intra, 0, nil)
+}
+
+func (g *generator) generateAssetTxnInternalHint(txType interface{}, sp sdk_types.SuggestedParams, intra uint64, hintIndex uint64, hint *assetData) (txn types.Transaction) {
+	// If there are no assets the next operation needs to be a create.
+	if len(g.assets) == 0 {
+		txType = assetCreate
+	}
+
+	var err error
+	numAssets := uint64(len(g.assets))
+	var senderIndex uint64
+
+	if txType == assetCreate {
+		senderIndex = numAssets % g.numAccounts
+		senderAcct := indexToAccount(senderIndex)
+		senderAcctStr := senderAcct.String()
+
+		total := assetTotal
+		assetID := g.txnCounter + intra + 1
+		assetName := fmt.Sprintf("asset #%d", assetID)
+		txn, err = future.MakeAssetCreateTxn(senderAcctStr, nil, sp, total, 0, false, senderAcctStr, senderAcctStr, senderAcctStr, senderAcctStr, "tokens", assetName, "https://algorand.com", "metadata!")
+		util.MaybeFail(err, "unable to make asset create transaction")
+
+		// Compute asset ID and initialize holdings
+		a := assetData{
+			assetID:  assetID,
+			creator:  senderIndex,
+			holdings: []assetHolding{{acctIndex: senderIndex, balance: total}},
+			holders:  map[uint64]bool{senderIndex: true},
+		}
+
+		g.assets = append(g.assets, &a)
+	} else {
+		assetIndex := rand.Uint64() % numAssets
+		asset := g.assets[assetIndex]
+		if hint != nil {
+			assetIndex = hintIndex
+			asset = hint
+		}
+
+		switch txType {
+		case assetDestroy:
+			// delete asset
+
+			// If the creator doesn't have all of them, close instead
+			if asset.holdings[0].balance != assetTotal {
+				return g.generateAssetTxnInternalHint(assetClose, sp, intra, assetIndex, asset)
+			}
+
+			senderIndex = asset.creator
+			creator := indexToAccount(senderIndex)
+			creatorString := creator.String()
+
+			txn, err = future.MakeAssetDestroyTxn(creatorString, nil, sp, asset.assetID)
+			util.MaybeFail(err, "unable to make asset destroy transaction")
+
+			// Remove asset by swapping the element to delete then trimming the last
+			g.assets[numAssets-1], g.assets[assetIndex] = g.assets[assetIndex], g.assets[numAssets-1]
+			g.assets = g.assets[:numAssets-1]
+		case assetOptin:
+			// select a random account from asset to optin
+
+			// If every account holds the asset, close instead of optin
+			if uint64(len(asset.holdings)) == g.numAccounts {
+				return g.generateAssetTxnInternalHint(assetClose, sp, intra, assetIndex, asset)
+			}
+
+			// look for an account that does not hold the asset
+			exists := true
+			for exists {
```

Comment:
No, you need the postcondition so it would look like this:
```
for senderIndex = rand.Uint64() % g.numAccounts; asset.holders[senderIndex]; senderIndex = rand.Uint64() % g.numAccounts {}
```

A rare example where do-while is the right loop:
```
do {
    senderIndex = rand.Uint64() % g.numAccounts
} while asset.holders[senderIndex]